### DISCUSSION TITLE
4. Cache System

### DISCUSSION THREAD

**Discussion by pysan3** [nvim-neo-tree/neo-tree.nvim#1338] [Internal Development]
### 5. Cache System

The current implementation (eg filesystem) will gather file info after `source.navigate` and then render the whole content from top to bottom.
I want to experiment if using `uv.new_fs_event` and a coroutine system (eg [nvim-nio](https://github.com/nvim-neotest/nvim-nio)) might make it possible to pre-calculate the file tree in the background and drastically improve the rendering speed after `:Neotree foo bar` is called.

**Comment by cseickel** [comment]
I think this is an over optimization, and Nui might already be doing this work to only re-render the changed portion.

If you just mean the scanning of the filesystem itself, yes that can be cached so we only re-scan the folders that have changed and make an incremental update to an internal representation of the file system. The code paths used for lazy loading folders could be utilized for this purpose.

**Reply by pysan3** [reply]
Yes, I'm only thinking of caching the file tree (so `buffers, doc_symbols etc` will get no changes).

According to this, I am thinking of making `use_libuv_file_watcher` default to true, as this should drastically improve the startup performance (time it takes to show the file tree after executing `:Neotree left`) if implemented correctly.

**Reply by cseickel** [reply]
I can see making the filewatcher on by default for the next major version, but we still have to accommodate those that turn it off.

I suppose that when it is on, we can just mark folders as stale if they are changed while the tree is closed, and then when it re-open we only have to refresh those stale folders which is more efficient in a large code base.
